### PR TITLE
Notifications api

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
Updates notifications api for connection manager.

### CODE REVIEW

File: libsplinter/src/network/connection_manager/mod.rs [hunk 1]
```diff
@@ -207,10 +207,11 @@ pub struct NotificationHandler {
 }
 
 impl NotificationHandler {
-    pub fn listen(&self) -> Result<CmNotification, ConnectionManagerError> {
-        match self.recv.recv() {
-            Ok(notifications) => Ok(notifications),
-            Err(_) => Err(ConnectionManagerError::SendMessageError(
+    pub fn try_next(&self) -> Result<Option<CmNotification>, ConnectionManagerError> {
+        match self.recv.try_recv() {
```
Comment (peterschwarz) [inline libsplinter/src/network/connection_manager/mod.rs]: What's the justification for this? There's not information in the commit message explaining why this is the case.

It seems like it may just introduce busy-wait situation for the user. Shouldn't there still be the blocking version of this method?

File: libsplinter/src/network/connection_manager/mod.rs [hunk 2]
```diff
@@ -227,6 +227,20 @@ impl NotificationHandler {
...
 }
 
+impl Iterator for NotificationHandler {
+    type Item = CmNotification;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.recv.recv() {
+            Ok(notification) => Some(notification),
+            Err(_) => {
+                error!("connection manager no longer running");
```
Comment (peterschwarz) [inline libsplinter/src/network/connection_manager/mod.rs]: Is this really an error? This message will be displayed on a clean shutdown. Probably should be debug message.

File: libsplinter/src/network/connection_manager/messages.rs [hunk 3]
```diff
@@ -18,8 +18,7 @@ use crate::network::connection_manager::error::ConnectionManagerError;
 
 pub enum CmMessage {
     Shutdown,
-    Subscribe(String, SyncSender<CmNotification>),
```
Comment (peterschwarz) [inline libsplinter/src/network/connection_manager/messages.rs]: Could you add some context to the commit message? I'm not sure I understand why this change is being made.

File: libsplinter/src/network/connection_manager/mod.rs [hunk 4]
```diff
@@ -227,6 +227,20 @@ impl NotificationHandler {
     }
 }
 
+impl Iterator for NotificationHandler {
```
Comment (peterschwarz) [inline libsplinter/src/network/connection_manager/mod.rs]: I think the more appropriate trait for this use case is to implement `IntoIterator`.  See https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#impl-IntoIterator.  It should be pretty simple, as you can just return the Receiver.

File: libsplinter/src/network/connection_manager/mod.rs (hunk 1 above)
Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: Updated commit message

File: libsplinter/src/network/connection_manager/mod.rs [hunk 5]
```diff
@@ -227,6 +227,20 @@ impl NotificationHandler {
...
+    type Item = CmNotification;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.recv.recv() {
+            Ok(notification) => Some(notification),
+            Err(_) => {
+                warn!("connection manager no longer running");
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: Need a more accurate error message here, and we shouldn't throw away the root cause.

File: libsplinter/src/network/connection_manager/mod.rs [hunk 6]
```diff
@@ -686,4 +700,50 @@ pub mod tests {
...
+    /// creates a thread that send HeartbeatSent notifications to
+    /// the notifier.
+    ///
+    /// Asserts:
+    ///
+    /// The notifications sent are received by the Notifier
+    /// correctly
+    ///
+    /// That the total number of notifications sent equals 5
+    fn test_nofication_handler_iterator() {
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: Typo - nofication

File: libsplinter/src/network/connection_manager/mod.rs [hunk 7]
```diff
@@ -361,29 +340,27 @@ fn handle_request(req: CmRequest, state: &mut ConnectionState) {
...
 fn notify_subscribers(
-    subscribers: &mut HashMap<String, SyncSender<CmNotification>>,
+    subscribers: &mut Vec<SyncSender<CmNotification>>,
     notification: CmNotification,
 ) {
-    for (id, sender) in subscribers.clone() {
-        if let Err(_) = sender.send(notification.clone()) {
+    subscribers.retain(|sender| {
+        if sender.send(notification.clone()).is_err() {
             warn!("subscriber has dropped its connection to connection manager");
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: Do we need some sort of id for this to be useful? How would we use this message when problem solving?

File: libsplinter/src/network/connection_manager/mod.rs [hunk 8]
```diff
@@ -227,6 +227,20 @@ impl NotificationHandler {
     }
 }
 
+impl Iterator for NotificationHandler {
+    type Item = CmNotification;
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: This should probably become ConnectionManagerNotification or just Notification as part of the public API (Cm being for private implementation).

File: libsplinter/src/network/connection_manager/mod.rs (hunk 5 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs]: I don't think this is the correct spot to log this as an error condition.  The only error returned is that the channel disconnected.  That could happen due to a clean shutdown or an error condition of the sending end, neither of which can be properly debugged by a message here.  This should be a debug message (as previously suggested), so that we know the receiver has exited, but any error condition should be caught elsewhere.

File: libsplinter/src/network/connection_manager/mod.rs (hunk 1 above)
Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: This is now hidden behind a feature

File: libsplinter/src/network/connection_manager/mod.rs (hunk 4 above)
Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: Could you elaborate a bit? I'm sure why this would be more appropriate.

File: libsplinter/src/network/connection_manager/messages.rs (hunk 3 above)
Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/messages.rs]: Updated commit message. Unsubscribe is unnecessary because you'd be using notifier like an iterator and calling unsubscribe on an iterator makes not sense, since it'll stop iterating when the sender goes out of scope.

File: libsplinter/src/network/connection_manager/mod.rs (hunk 7 above)
Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: Hmm maybe this shouldn't be `warn!` since it's completely valid for the subscribe to drop out of scope.

File: libsplinter/src/network/connection_manager/mod.rs (hunk 4 above)
Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs]: I guess it's a question of whether or not a NotificationHandler should _be_ an iterator or can be transformed _into_ an iterator.

Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: I think it makes more sense as an Iterator, since it doesn't have much purpose otherwise, and having it implement `IntoIterator` instead just seems to add an extra step (calling `notifier.next()` vs calling `notifier.into_iter().next()`.

Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs]: I'm not sure I agree with that usage, though.  By hiding the blocking version behind `Iterator::next`, it's not exactly documented.  It should be its own method on Notifier, and the iterator provide just calls it.

Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: Ok. So you're saying don't remove `listen`? Because what you're describing is what `listen` was. Should we even have an iterator then?

Comment (peterschwarz) [reply libsplinter/src/network/connection_manager/mod.rs]: We can still have an iterator, but `listen` should be part of the public API.  Iterator is still a nice convenience, so it can be both.

Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: Changed `try_next` to `try_listen`, and add listen back.

File: libsplinter/src/network/connection_manager/mod.rs [hunk 9]
```diff
@@ -216,6 +218,19 @@ impl NotificationHandler {
         }
     }
 
+    #[cfg(feature = "connection-manager-notifier-try-next")]
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: The name of this feature doesn't match the function (try_listen) and try_listen doesn't make as much as much sense as calling it try_next.

File: libsplinter/src/network/connection_manager/mod.rs [hunk 10]
```diff
@@ -216,17 +209,34 @@ impl NotificationHandler {
...
     }
 }
 
+impl Iterator for Notifier {
+    type Item = ConnectionManagerNotification;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.recv.recv() {
+            Ok(notification) => Some(notification),
+            Err(_) => {
```
Comment (vaporos) [inline libsplinter/src/network/connection_manager/mod.rs]: I think the Notifier should explicitly recv a end-of-iteration marker so that if we get an err on receive it is unexpected and means there was a coding error.

Comment (RyanLassigBanks) [reply libsplinter/src/network/connection_manager/mod.rs]: I'm not entirely sure how this would work, or if it's necessary. Again, `recv` can only "fail" if the connection manger has gone away either by shutting down gracefully or by panicking. Regardless, from the perspective of the iterator, it's done iterating. Adding an explicit marker seems redundant since the "marker" is the sender end going out of scope.

It's worth noting there is a `FatalError` notification that gets sent out if the connection manager is forced to shutdown unexpectedly, so the subscriber would be notified if something happened to connection manager and would be able to act accordingly.

Comment (vaporos) [reply libsplinter/src/network/connection_manager/mod.rs]: Ok, suggest removing the debug! then (completely uninteresting from a debugging standpoint, as you can't tell if it is normal operation or not) and adding a comment on that line reading "This is expected if the other side shuts down before this end." (re-used from Transact code).
